src/transaction/log_recovery_redo_parallel.hpp

```diff
@@ -0,0 +1,274 @@
+  */
+  class redo_parallel::redo_job_base
+  {
+    public:
+      redo_job_base (VPID a_vpid)
+	: m_vpid (a_vpid)
+      {
+	// the logic implemented below, in the queue and task, does allow for null-vpid type of
+	// entries to be executed - they are called "to be waited for" or "synched" operations;
+	// but, for now, do not allow them to be dispatched to be executed asynchronously
+	assert (!VPID_ISNULL (&m_vpid));
+      }
+      redo_job_base () = delete;
+      redo_job_base (redo_job_base const &) = delete;
+      redo_job_base (redo_job_base &&) = delete;
+
+      virtual ~redo_job_base () = default;
+
+      redo_job_base &operator = (redo_job_base const &) = delete;
+      redo_job_base &operator = (redo_job_base &&) = delete;
+
+      /* log entries come in more than one flavor:
+      *  - pertain to a certain vpid - aka: page update
+      *  - pertain to a certain VOLID - aka: volume extend, new page
+      *  - pertain to no VOLID - aka: database extend, new volume
+      */
+      const VPID &get_vpid () const
+      {
+	return m_vpid;
+      }
+
+      virtual int execute (THREAD_ENTRY *thread_p, log_reader &log_pgptr_reader,
+			   LOG_ZIP &undo_unzip_support, LOG_ZIP &redo_unzip_support) = 0;
+
+    private:
+      const VPID m_vpid;
+  };
+
+
+  /* actual job implementation that performs log recovery redo
+   */
+  template <typename TYPE_LOG_REC>
+  class redo_job_impl final : public redo_parallel::redo_job_base
+  {
+      using log_rec_t = TYPE_LOG_REC;
+
+    public:
+      redo_job_impl () = delete;
+      redo_job_impl (VPID a_vpid, const log_lsa &a_rcv_lsa, const LOG_LSA *a_end_redo_lsa, LOG_RECTYPE a_log_rtype)
+	: redo_parallel::redo_job_base (a_vpid)
+	, m_rcv_lsa (a_rcv_lsa)
+	, m_end_redo_lsa (a_end_redo_lsa)
+	, m_log_rtype (a_log_rtype)
+      {
+      }
+
+      redo_job_impl (redo_job_impl const &) = delete;
+      redo_job_impl (redo_job_impl &&) = delete;
+
+      ~redo_job_impl () override = default;
+
+      redo_job_impl &operator = (redo_job_impl const &) = delete;
+      redo_job_impl &operator = (redo_job_impl &&) = delete;
+
+      int execute (THREAD_ENTRY *thread_p, log_reader &log_pgptr_reader,
+		   LOG_ZIP &undo_unzip_support, LOG_ZIP &redo_unzip_support) override
+      {
+	const int err_set_lsa_and_fetch_page =  log_pgptr_reader.set_lsa_and_fetch_page (m_rcv_lsa);
+	if (err_set_lsa_and_fetch_page != NO_ERROR)
+	  {
+	    return err_set_lsa_and_fetch_page;
+	  }
+	log_pgptr_reader.add_align (sizeof (LOG_RECORD_HEADER));
+	log_pgptr_reader.advance_when_does_not_fit (sizeof (log_rec_t));
+	const log_rec_t log_rec
+	  = log_pgptr_reader.reinterpret_copy_and_add_align<log_rec_t> ();
+
+	const auto &rcv_vpid = get_vpid ();
+	log_rv_redo_record_sync<log_rec_t> (thread_p, log_pgptr_reader, log_rec, rcv_vpid, m_rcv_lsa, m_end_redo_lsa,
+					    m_log_rtype, undo_unzip_support, redo_unzip_support);
+	return NO_ERROR;
+      }
```

Comment:
Move definition to .cpp file.

Reply:
Tried.
Linker error because the template is instantiated externally in `log_recovery.c`:
```      
    using redo_job_impl_t = cublog::redo_job_impl<T>;
    std::unique_ptr<redo_job_impl_t> job{ new redo_job_impl_t (rcv_vpid, rcv_lsa, end_redo_lsa, log_rtype) };
```